cli: COCKROACH_URL env var is broken

**Describe the problem**

The `COCKROACH_URL` environment variable is supposed to be equivalent to the `--url` flag, but using it causes a panic.

**To Reproduce**

`cockroach sql --url` works:

```
[bdarnell@ben-crl ~]$ cockroach sql --url=postgresql://root@127.0.0.1:26257?sslmode=disable
# Welcome to the cockroach SQL interface.
# All statements must be terminated by a semicolon.
# To exit: CTRL + D.
#
# Server version: CockroachDB CCL v19.1.4 (x86_64-apple-darwin14, built 2019/08/06 15:29:00, go1.11.6) (same version as client)
# Cluster ID: 06389009-5b11-4adb-97dc-a52b72a0f0d5
#
# Enter \? for a brief introduction.
#
root@127.0.0.1:26257/defaultdb> ^D   
```

`cockroach sql` with a `COCKROACH_URL` env var panics:
```                                                                                                                                          [bdarnell@ben-crl ~]$ COCKROACH_URL=postgresql://root@127.0.0.1:26257?sslmode=disable cockroach sql
warning: --url specifies user/password, but command "sql" does not accept user/password details - details ignored
panic: invalid argument "postgresql://root@127.0.0.1:26257?sslmode=disable" for "--url" flag: extracting host/port: no such flag -host

goroutine 1 [running]:
github.com/cockroachdb/cockroach/pkg/cli.setFlagFromEnv(0xc00082a200, 0x6afda86, 0x3, 0x0, 0x0, 0x6b1035e, 0xd, 0x6bd480d, 0xa1)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/flags.go:91 +0xa0
github.com/cockroachdb/cockroach/pkg/cli.VarFlag(0xc00082a200, 0x727fbc0, 0xc00036a860, 0x6afda86, 0x3, 0x0, 0x0, 0x6b1035e, 0xd, 0x6bd480d, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/flags.go:131 +0xfc
github.com/cockroachdb/cockroach/pkg/cli.init.5()
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/flags.go:507 +0x29c4
```

**Expected behavior**

The env var should work in the same way as the command-line flag.


**Environment:**
 - CockroachDB version 19.1.4
 - Server OS: macOS
 - Client app `cockroach sql`

**Additional context**

I think we should be encouraging the use of `COCKROACH_URL` in documentation, but it has to work first.

--- a/pkg/cli/client_url.go
+++ b/pkg/cli/client_url.go
@@ -97,12 +97,13 @@ func (u urlParser) Set(v string) error {
 	}
 
 	cliCtx := u.cliCtx
+	fl := flagSetForCmd(u.cmd)
 
 	// If user name / password information is available, forward it to
 	// --user. We store the password for later re-collection by
 	// makeClientConnURL().
 	if parsedURL.User != nil {
-		f := u.cmd.Flags().Lookup(cliflags.User.Name)
+		f := fl.Lookup(cliflags.User.Name)
 		if f == nil {
 			// A client which does not support --user will also not use
 			// makeClientConnURL(), so we can ignore/forget about the
@@ -141,7 +142,7 @@ func (u urlParser) Set(v string) error {
 	// If a database path is available, forward it to --database.
 	if parsedURL.Path != "" {
 		dbPath := strings.TrimLeft(parsedURL.Path, "/")
-		f := u.cmd.Flags().Lookup(cliflags.Database.Name)
+		f := fl.Lookup(cliflags.Database.Name)
 		if f == nil {
 			// A client which does not support --database does not need this
 			// bit of information, so we can ignore/forget about it. We do
@@ -168,8 +169,6 @@ func (u urlParser) Set(v string) error {
 		}
 
 		cliCtx.extraConnURLOptions = options
-
-		fl := u.cmd.Flags()
 
 		switch sslMode := options.Get("sslmode"); sslMode {
 		case "disable":

--- a/pkg/cli/demo.go
+++ b/pkg/cli/demo.go
@@ -140,8 +140,9 @@ func setupTransientServers(
 
 	// Set up logging. For demo/transient server we use non-standard
 	// behavior where we avoid file creation if possible.
-	df := cmd.Flags().Lookup(cliflags.LogDir.Name)
-	sf := cmd.Flags().Lookup(logflags.LogToStderrName)
+	fl := flagSetForCmd(cmd)
+	df := fl.Lookup(cliflags.LogDir.Name)
+	sf := fl.Lookup(logflags.LogToStderrName)
 	if !df.Changed && !sf.Changed {
 		// User did not request logging flags; shut down all logging.
 		// Otherwise, the demo command would cause a cockroach-data
@@ -374,7 +375,7 @@ func runDemo(cmd *cobra.Command, gen workload.Generator) error {
 	}
 
 	// If the geo-partitioned replicas flag was given and the nodes have changed, throw an error.
-	if demoCtx.geoPartitionedReplicas && cmd.Flags().Lookup(cliflags.DemoNodes.Name).Changed {
+	if demoCtx.geoPartitionedReplicas && flagSetForCmd(cmd).Lookup(cliflags.DemoNodes.Name).Changed {
 		return errors.New("--nodes cannot be used with --geo-partitioned-replicas")
 	}
 

--- a/pkg/cli/flags.go
+++ b/pkg/cli/flags.go
@@ -229,6 +229,14 @@ const maxClusterNameLength = 256
 
 const backgroundEnvVar = "COCKROACH_BACKGROUND_RESTART"
 
+// flagSetForCmd is a replacement for cmd.Flag() that properly merges
+// persistent and local flags, until the upstream bug
+// https://github.com/spf13/cobra/issues/961 has been fixed.
+func flagSetForCmd(cmd *cobra.Command) *pflag.FlagSet {
+	_ = cmd.LocalFlags() // force merge persistent+local flags
+	return cmd.Flags()
+}
+
 func init() {
 	initCLIDefaults()
 
@@ -555,7 +563,7 @@ func init() {
 	// Make the other non-SQL client commands also recognize --url in
 	// strict SSL mode.
 	for _, cmd := range clientCmds {
-		if f := cmd.Flags().Lookup(cliflags.URL.Name); f != nil {
+		if f := flagSetForCmd(cmd).Lookup(cliflags.URL.Name); f != nil {
 			// --url already registered above, nothing to do.
 			continue
 		}
@@ -645,11 +653,11 @@ func extraServerFlagInit(cmd *cobra.Command) error {
 		serverSQLPort = serverListenPort
 	}
 	serverCfg.SQLAddr = net.JoinHostPort(serverSQLAddr, serverSQLPort)
-	serverCfg.SplitListenSQL = cmd.Flags().Lookup(cliflags.ListenSQLAddr.Name).Changed
+	serverCfg.SplitListenSQL = flagSetForCmd(cmd).Lookup(cliflags.ListenSQLAddr.Name).Changed
 
 	// Fill in the defaults for --advertise-sql-addr.
-	advSpecified := cmd.Flags().Lookup(cliflags.AdvertiseAddr.Name).Changed ||
-		cmd.Flags().Lookup(cliflags.AdvertiseHost.Name).Changed
+	advSpecified := flagSetForCmd(cmd).Lookup(cliflags.AdvertiseAddr.Name).Changed ||
+		flagSetForCmd(cmd).Lookup(cliflags.AdvertiseHost.Name).Changed
 	if serverSQLAdvertiseAddr == "" {
 		if advSpecified {
 			serverSQLAdvertiseAddr = serverAdvertiseAddr
@@ -704,9 +712,7 @@ func extraClientFlagInit() {
 }
 
 func setDefaultStderrVerbosity(cmd *cobra.Command, defaultSeverity log.Severity) error {
-	pf := cmd.Flags()
-
-	vf := pf.Lookup(logflags.LogToStderrName)
+	vf := flagSetForCmd(cmd).Lookup(logflags.LogToStderrName)
 
 	// if `--logtostderr` was not specified and no log directory was
 	// set, or `--logtostderr` was specified but without explicit level,

--- a/pkg/cli/flags.go
+++ b/pkg/cli/flags.go
@@ -90,35 +90,25 @@ func AddPersistentPreRunE(cmd *cobra.Command, fn func(*cobra.Command, []string) 
 	}
 }
 
-func setFlagFromEnv(f *pflag.FlagSet, flagInfo cliflags.FlagInfo) {
-	if flagInfo.EnvVar != "" {
-		if value, set := envutil.EnvString(flagInfo.EnvVar, 2); set {
-			if err := f.Set(flagInfo.Name, value); err != nil {
-				panic(err)
-			}
-		}
-	}
-}
-
 // StringFlag creates a string flag and registers it with the FlagSet.
 func StringFlag(f *pflag.FlagSet, valPtr *string, flagInfo cliflags.FlagInfo, defaultVal string) {
 	f.StringVarP(valPtr, flagInfo.Name, flagInfo.Shorthand, defaultVal, flagInfo.Usage())
 
-	setFlagFromEnv(f, flagInfo)
+	registerEnvVarDefault(f, flagInfo)
 }
 
 // IntFlag creates an int flag and registers it with the FlagSet.
 func IntFlag(f *pflag.FlagSet, valPtr *int, flagInfo cliflags.FlagInfo, defaultVal int) {
 	f.IntVarP(valPtr, flagInfo.Name, flagInfo.Shorthand, defaultVal, flagInfo.Usage())
 
-	setFlagFromEnv(f, flagInfo)
+	registerEnvVarDefault(f, flagInfo)
 }
 
 // BoolFlag creates a bool flag and registers it with the FlagSet.
 func BoolFlag(f *pflag.FlagSet, valPtr *bool, flagInfo cliflags.FlagInfo, defaultVal bool) {
 	f.BoolVarP(valPtr, flagInfo.Name, flagInfo.Shorthand, defaultVal, flagInfo.Usage())
 
-	setFlagFromEnv(f, flagInfo)
+	registerEnvVarDefault(f, flagInfo)
 }
 
 // DurationFlag creates a duration flag and registers it with the FlagSet.
@@ -127,14 +117,14 @@ func DurationFlag(
 ) {
 	f.DurationVarP(valPtr, flagInfo.Name, flagInfo.Shorthand, defaultVal, flagInfo.Usage())
 
-	setFlagFromEnv(f, flagInfo)
+	registerEnvVarDefault(f, flagInfo)
 }
 
 // VarFlag creates a custom-variable flag and registers it with the FlagSet.
 func VarFlag(f *pflag.FlagSet, value pflag.Value, flagInfo cliflags.FlagInfo) {
 	f.VarP(value, flagInfo.Name, flagInfo.Shorthand, flagInfo.Usage())
 
-	setFlagFromEnv(f, flagInfo)
+	registerEnvVarDefault(f, flagInfo)
 }
 
 // StringSlice creates a string slice flag and registers it with the FlagSet.
@@ -142,7 +132,7 @@ func StringSlice(
 	f *pflag.FlagSet, valPtr *[]string, flagInfo cliflags.FlagInfo, defaultVal []string,
 ) {
 	f.StringSliceVar(valPtr, flagInfo.Name, defaultVal, flagInfo.Usage())
-	setFlagFromEnv(f, flagInfo)
+	registerEnvVarDefault(f, flagInfo)
 }
 
 // aliasStrVar wraps a string configuration option and is meant
@@ -239,6 +229,11 @@ func flagSetForCmd(cmd *cobra.Command) *pflag.FlagSet {
 
 func init() {
 	initCLIDefaults()
+	defer func() {
+		if err := processEnvVarDefaults(); err != nil {
+			panic(err)
+		}
+	}()
 
 	// Every command but start will inherit the following setting.
 	AddPersistentPreRunE(cockroachCmd, func(cmd *cobra.Command, _ []string) error {
@@ -629,6 +624,51 @@ func init() {
 	{
 		f := debugBallastCmd.Flags()
 		VarFlag(f, &debugCtx.ballastSize, cliflags.Size)
+	}
+}
+
+// processEnvVarDefaults injects the current value of flag-related
+// environment variables into the initial value of the settings linked
+// to the flags, during initialization and before the command line is
+// actually parsed. For example, it will inject the value of
+// $COCKROACH_URL into the urlParser object linked to the --url flag.
+func processEnvVarDefaults() error {
+	for envVar, d := range envVarDefaults {
+		if err := d.flagSet.Set(d.flagName, d.envValue); err != nil {
+			return errors.Wrapf(err, "setting --%s from %s", d.flagName, envVar)
+		}
+	}
+	return nil
+}
+
+// envVarDefault describes a delayed default initialization of the
+// setting covered by a flag from the value of an environment
+// variable.
+type envVarDefault struct {
+	envValue string
+	flagName string
+	flagSet  *pflag.FlagSet
+}
+
+// envVarDefaults records the initializations from environment variables
+// for processing at the end of initialization, before flag parsing.
+var envVarDefaults = map[string]envVarDefault{}
+
+// registerEnvVarDefault registers a deferred initialization of a flag
+// from an environment variable.
+func registerEnvVarDefault(f *pflag.FlagSet, flagInfo cliflags.FlagInfo) {
+	if flagInfo.EnvVar == "" {
+		return
+	}
+	value, set := envutil.EnvString(flagInfo.EnvVar, 2)
+	if !set {
+		// Env var not set. Nothing to do.
+		return
+	}
+	envVarDefaults[flagInfo.EnvVar] = envVarDefault{
+		envValue: value,
+		flagName: flagInfo.Name,
+		flagSet:  f,
 	}
 }
 

--- a/pkg/cli/interactive_tests/common.tcl
+++ b/pkg/cli/interactive_tests/common.tcl
@@ -89,7 +89,7 @@ proc send_eof {} {
 # in `server_pid`.
 proc start_server {argv} {
     report "BEGIN START SERVER"
-    system "$argv start-single-node --insecure --pid-file=server_pid --background -s=path=logs/db >>logs/expect-cmd.log 2>&1;
+    system "$argv start-single-node --insecure --pid-file=server_pid --listening-url-file=server_url --background -s=path=logs/db >>logs/expect-cmd.log 2>&1;
             $argv sql --insecure -e 'select 1'"
     report "START SERVER DONE"
 }

--- a/pkg/cli/interactive_tests/test_flags.tcl
+++ b/pkg/cli/interactive_tests/test_flags.tcl
@@ -80,5 +80,19 @@ interrupt
 eexpect ":/# "
 end_test
 
+
+start_server $argv
+
+start_test "Check that a client can connect using the URL env var"
+send "export COCKROACH_URL=`cat server_url`;\r"
+eexpect ":/# "
+send "$argv sql\r"
+eexpect "defaultdb>"
+interrupt
+eexpect ":/# "
+end_test
+
+stop_server $argv
+
 send "exit 0\r"
 eexpect eof

--- a/pkg/cli/start.go
+++ b/pkg/cli/start.go
@@ -414,7 +414,7 @@ func initTempStorageConfig(
 var errCannotUseJoin = errors.New("cannot use --join with 'cockroach start-single-node' -- use 'cockrach start' instead")
 
 func runStartSingleNode(cmd *cobra.Command, args []string) error {
-	joinFlag := cmd.Flags().Lookup(cliflags.Join.Name)
+	joinFlag := flagSetForCmd(cmd).Lookup(cliflags.Join.Name)
 	if joinFlag.Changed {
 		return errCannotUseJoin
 	}
@@ -501,8 +501,7 @@ func runStart(cmd *cobra.Command, args []string, disableReplication bool) error 
 	grpcutil.SetSeverity(log.Severity_WARNING)
 
 	// Check the --join flag.
-	pf := cmd.Flags()
-	if !pf.Lookup(cliflags.Join.Name).Changed {
+	if !flagSetForCmd(cmd).Lookup(cliflags.Join.Name).Changed {
 		log.Shout(ctx, log.Severity_WARNING,
 			"running 'cockroach start' without --join is deprecated.\n"+
 				"Consider using 'cockroach start-single-node' or 'cockroach init' instead.")
@@ -981,7 +980,7 @@ If problems persist, please see ` + base.DocsURL("cluster-setup-troubleshooting.
 }
 
 func hintServerCmdFlags(ctx context.Context, cmd *cobra.Command) {
-	pf := cmd.Flags()
+	pf := flagSetForCmd(cmd)
 
 	listenAddrSpecified := pf.Lookup(cliflags.ListenAddr.Name).Changed || pf.Lookup(cliflags.ServerHost.Name).Changed
 	advAddrSpecified := pf.Lookup(cliflags.AdvertiseAddr.Name).Changed || pf.Lookup(cliflags.AdvertiseHost.Name).Changed

